unknown properties in schema when set timestams to true

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

6.8.0

### Node.js version

16.15.1

### MongoDB server version

6.0

### Typescript version (if applicable)

4.9.4

### Description

In version 6.8.0, when I set the timestamps option to true in a schema, the created model has properties with unknown types that will cause a lot of problems in implementing the application. However, when I switch back to version 6.7.5 the problem disappears.
.
.
**Version 6.8.0:**
```
(property) Users: Model<{
    createdAt: NativeDate;
    updatedAt: NativeDate;
} & {
    email?: unknown;
    password?: unknown;
    name?: unknown;
}, {}, {
    // This is a custom method in the schema
    comparePassword: (this: Document<unknown, any, FlatRecord<{
        email: string;
        password: string;
        name: string;
...
```
.
.
**When switched back to version 6.7.5 without any modification:**
```
(property) Users: Model<{
    email: string;
    password: string;
    name: string;
}, {}, {
    comparePassword: (this: Document<unknown, any, FlatRecord<{
        email: string;
        password: string;
        name: string;
...
```

### Steps to Reproduce

Considering you are using typescript:
1. Install version 6.8.0 : npm i mongoose@6.8.0
2. Create a schema with some properties and set the timestamps option to true.
3. Create a connection using mongoose.createConnection.
4. Create a model for the schema using the connection.model.
5. Hover over the created model to see the type of the model.

### Expected Behavior

**Expected behavior:**
```
(property) Users: Model<{
    createdAt: NativeDate;
    updatedAt: NativeDate;
} & {
    email: string;
    password: string;
    name: string;
}, {}, {
    // This is a custom method in the schema
    comparePassword: (this: Document<unknown, any, FlatRecord<{
        email: string;
        password: string;
        name: string;
...
```
Or:
```
(property) Users: Model<{
    email: string;
    password: string;
    name: string;
}, {}, {
    // This is a custom method in the schema
    comparePassword: (this: Document<unknown, any, FlatRecord<{
        email: string;
        password: string;
        name: string;
...
```

Confirmed that the following script fails to compile due to `gh-12807.ts:46:9 - error TS2322: Type 'unknown' is not assignable to type 'string'.`

```ts
import { Schema, createConnection } from 'mongoose';

const userSchema = new Schema(
  {
    email: {
      type: String,
      required: true,
    },

    password: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
  },
  {
    methods: {
      comparePassword: function (candidatePassword: string): Promise<boolean> {
        const user = this;

        return new Promise((resolve, reject) => {
          resolve(true);
        });
      },
    },

    timestamps: true,
  }
);

const connection = createConnection('mongodb://localhost:27017/test');

const Models = {
  Users: connection.model("users", userSchema),
};

const user = new Models.Users({ name: 'foo', password: 'bar', email: 'baz' });

const name: string = user.name;

async function foo() {
  const user = await Models.Users.findOne().orFail();
  const name: string = user.name;
}
```

--- a/docs/typescript/schemas.md
+++ b/docs/typescript/schemas.md
@@ -30,7 +30,7 @@ There are a few caveats for using automatic type inference:
 
 1. You need to set `strictNullChecks: true` or `strict: true` in your `tsconfig.json`. Or, if you're setting flags at the command line, `--strictNullChecks` or `--strict`. There are [known issues](https://github.com/Automattic/mongoose/issues/12420) with automatic type inference with strict mode disabled.
 2. You need to define your schema in the `new Schema()` call. Don't assign your schema definition to a temporary variable. Doing something like `const schemaDefinition = { name: String }; const schema = new Schema(schemaDefinition);` will not work.
-3. Mongoose adds `createdAt` and `updatedAt` to your schema if you specify the `timestamps` option in your schema, *except* if you also specify `methods`, `virtuals`, or `statics`. There is a [known issue](https://github.com/Automattic/mongoose/issues/12807) with type inference with timestamps and methods/virtuals/statics options. If you use methods, virtuals, and statics, you're responsible for adding `createdAt` and `updatedAt` to your schema definition.
+3. Mongoose adds `createdAt` and `updatedAt` to your schema if you specify the `timestamps` option in your schema.
 
 If you must define your schema separately, use [as const](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) (`const schemaDefinition = { ... } as const;`) to prevent *type widening*. TypeScript will automatically widen types like `required: false` to `required: boolean`, which will cause Mongoose to assume the field is required. Using `as const` forces TypeScript to retain these types.
 

--- a/test/types/schema.test.ts
+++ b/test/types/schema.test.ts
@@ -899,11 +899,13 @@ function testInferTimestamps() {
   });
 
   type WithTimestamps2 = InferSchemaType<typeof schema2>;
-  // For some reason, expectType<{ createdAt: Date, updatedAt: Date, name?: string }> throws
-  // an error "Parameter type { createdAt: Date; updatedAt: Date; name?: string | undefined; }
-  // is not identical to argument type { createdAt: NativeDate; updatedAt: NativeDate; } &
-  // { name?: string | undefined; }"
-  expectType<{ name?: string | null }>({} as WithTimestamps2);
+  expectType<{ createdAt: Date; updatedAt: Date } & { name?: string | null }>({} as WithTimestamps2);
+
+  const TestModel = model('Test', schema2);
+  const doc = new TestModel({ name: 'test' });
+  expectType<string | undefined | null>(doc.name);
+
+  expectType<string | undefined | null>(doc.myName());
 }
 
 function gh12431() {

--- a/types/inferschematype.d.ts
+++ b/types/inferschematype.d.ts
@@ -75,10 +75,7 @@ declare module 'mongoose' {
 
   type ApplySchemaOptions<T, O = DefaultSchemaOptions> = ResolveTimestamps<T, O>;
 
-  type ResolveTimestamps<T, O> = O extends { methods: any } | { statics: any } | { virtuals: any } | { timestamps?: false } ? T
-    // For some reason, TypeScript sets all the document properties to unknown
-    // if we use methods, statics, or virtuals. So avoid inferring timestamps
-    // if any of these are set for now. See gh-12807
+  type ResolveTimestamps<T, O> = O extends { timestamps?: false } ? T
     : O extends { timestamps: infer TimestampOptions } ? TimestampOptions extends true
       ? { createdAt: NativeDate; updatedAt: NativeDate; } & T
       : TimestampOptions extends SchemaTimestampsConfig

